### ISSUE TITLE
ensure_softmax is slow

### ISSUE THREAD

**Issue by E-Delabrouille** [opendilab/LightZero#383]
I tracked Stochastic MuZero's performances with py-spy, both for a personal study case and for the provided 2048 application case. I'll speak here only about the 2048 case, which is easily reproducible, but my observations are the same on both application cases.

Here are my results :
<img width="857" height="193" alt="Image" src="https://github.com/user-attachments/assets/b43e0e0b-3283-49cd-ad6b-6bd1da1397d2" />

The functions in which we spend the most time (that is, ~5 times more than the others) are :
- `process_nodes` from `lzero/mcts/tree_search/mcts_ctree_stochastic.py`. This comes at 99% from the `.detach().cpu().numpy()` commands. Perhaps that they could be optimized but I have no idea on that precise matter.
- `ensure_softmax` from `lzero/policy/scaling_transform.py` (and its brother `__call__`). **This** surprises me a lot.

`ensure_softmax` is very often used in all of LightZero's algorithms :

```python
def ensure_softmax(logits, dim=1):
    """
    Overview:
        Ensure that the input tensor is normalized along the specified dimension.
    Arguments:
         - logits (:obj:`torch.Tensor`): The input tensor.
        - dim (:obj:`int`): The dimension along which to normalize the input tensor.
    Returns:
        - output (:obj:`torch.Tensor`): The normalized tensor.
    """
    # Calculate the sum along the specified dimension (dim=1 in this case)
    sum_along_dim = logits.sum(dim=dim, keepdim=True)

    # Create a tensor of ones with the same shape as sum_along_dim
    ones_like_sum = torch.ones_like(sum_along_dim)
    
    # Check if the logits are already normalized (i.e., if the sum along the dimension is approximately 1)
    # torch.allclose checks if all elements of two tensors are close within a tolerance
    # atol (absolute tolerance) is set to a small value to allow for numerical precision issues
    is_normalized = torch.allclose(sum_along_dim, ones_like_sum, atol=1e-5)
    
    # If logits are not normalized, apply softmax along the specified dimension
    if not is_normalized:
        return torch.softmax(logits, dim=dim)
    else:
        # If logits are already normalized, return them as they are
        return logits
```

At the beginning of the training, it's common to see theses `sum_along_dim` with an order of magntude of 10e3 (weird, for logits). For a well-trained model these sum are in the tens.
*In both cases* it seems that the input logits have _almost never_ `sum_along_dim` values near 1. If we accept this as an universal truth and replace `ensure_softmax` consequally :

```python
def ensure_softmax(logits, dim=1):
    return torch.softmax(logits, dim=dim)
```

The new py-spy profile for the same execution looks like so :
<img width="857" height="241" alt="Image" src="https://github.com/user-attachments/assets/64b2d7df-4b4a-484f-bd5b-fcc1036e21c8" />

**`ensure_softmax` part of execution time plummets from a third of the total execution time to a hundredth.**

**Conclusion : the test of whether or not the input logits are already normalized is _extremely_ costly and seems rarely necessary. My solution of completely ignoring them is surely not correct from a therocial point of view, but a great deal of optimization can be digged up from there. Perhaps there are cases where we know for sure whether or not the normalization is needed, for which the costly test could be skipped ?**

**Comment by puyuan1996** [comment]
Thank you very much for your meticulous discovery and for conducting the py-spy performance analysis. Your test results are extremely valuable and closely align with our analysis. Since in most cases the logits being passed in are not normalized, we should bypass the costly normalization check and apply the softmax directly. We will submit a PR later to address this issue. Thank you again!

**Comment by puyuan1996** [comment]
Hello, we have refined the ensure_softmax function as shown in https://github.com/opendilab/LightZero/pull/389. We warmly welcome your contributions towards optimizing the search method in mcts_ctree_stochastic.py. A potential optimized version is available at https://poe.com/s/nGYGxcsCM9FKygp7Hkaa. We hope you can test it using the 2048 game or your custom environment to evaluate the changes in performance and efficiency. Thank you for your contribution!

**Comment by Firerozes** [comment]
Hi again @puyuan1996 
Giving a look at this optimized version of yours, I saw it's o3 written.

Have you already tested its correctness and only want speedup metrics, or is its correctness unkown yet ?

**Comment by puyuan1996** [comment]
Hello, yes, the suggestion came from o3 as an optimization proposal. We haven’t tested its correctness yet, as we’ve been busy with other matters recently. When we get some free time, we plan to optimize the function accordingly. Since you’re working with stochastic_muzero research, the speed optimization of this search method could be quite important. This suggestion is offered as a reference. Thanks again for your attention.
